src/main/java/bio/terra/service/filedata/flight/ingest/IngestBulkGcpStep.java

```diff
@@ -180,10 +180,20 @@ private List<FileModel> performIngest(
     // attempting re-copying files)
     // For the case where we use predictable file Ids, this is a noop
     if (!dataset.hasPredictableFileIds()) {
-      doFileCopy(workingMap, loadModels, fileIdsByPath, bucketResource).stream()
+      List<CopyResult> copyResults =
+          doFileCopy(workingMap, loadModels, fileIdsByPath, bucketResource);
+      copyResults.stream()
           .filter(r -> r.error() == null && r.fsFileInfo() != null)
           .map(CopyResult::fsFileInfo)
           .forEach(fsFileInfos::add);
+      List<String> failedTargetPaths =
+          copyResults.stream()
+              .filter(r -> r.fsFileInfo() == null)
+              .map(CopyResult::targetPath)
+              .toList();
```

Comment:
I see that this is a pattern followed elsewhere in this file, but restreaming the same collection multiple times will be less performant than only going through it once.  Since bulk mode is most likely used for large volume ingests, I think the performance concern is worth addressing.

In this case, I also think that the two streams obscure what's going on.  What do you think about something like this instead?  It's similar to how we [sort through a stream for processing Rawls workspaces](https://github.com/DataBiosphere/jade-data-repo/blob/5e19d8b5e87a68bb6c244c33c1efbe4e3e4f77b2/src/main/java/bio/terra/service/rawls/RawlsService.java#L49C1-L66).
```
List<String> failedTargetPaths = new ArrayList<>();
doFileCopy(workingMap, loadModels, fileIdsByPath, bucketResource).stream()
  .forEach(copyResult -> {
    var fsFileInfo = copyResult.fsFileInfo();
    if (copyResult.error() == null && fsFileInfo != null) {
      fsFileInfos.add(fsFileInfo);
    } else {
      failedTargetPaths.add(copyResult.targetPath());
    }
  });
```
(I'm not positive on the conditional I put in there -- should we be considering a target path a failure if it has an error, but has `fsFileInfo`?  Would such a thing even happen?)

Reply:
Oh this is much nicer! Thanks for the suggestion 🙂